lib/mix/lib/mix/task.ex

```diff
@@ -347,29 +358,46 @@ defmodule Mix.Task do
   task is invalid. See `get!/1` for more information.
   """
   @spec run(task_name, [any]) :: any
-  def run(task, args \\ [])
+  def run(task, args \\ []) do
+    do_run(task, args)
+  end
+
+  @doc """
+  Similar to run/2 but it runs the task in the specified list of children apps,
+  in the context of umbrella proyects.
+
+  However, if the task is not recursive (whose purpose is to be run in children
+  applications) it runs at the project root level (a warning message is shown
+  in this case).
+  """
+  @spec run_in_apps(task_name, [atom], [any]) :: any
+  def run_in_apps(task, apps, args \\ []) do
+    do_run(task, args, apps: apps)
+  end
 
-  def run(task, args) when is_atom(task) do
-    run(Atom.to_string(task), args)
+  defp do_run(task, args, options \\ [])
```

Comment:
We typically don't use default arguments for private functions. Instead we change all callers. Personally speaking, I would remove the `options` and call it `in_apps`. it is either `nil` or a `list`. This should simplify a bit the code below and the amount of state kept.